src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+        /// </summary>
+        /// <param name="assumedValidHeader">The header that is assumed to be valid.</param>
+        /// <param name="latestNewHeader">The last header in the list of presented new headers.</param>
+        private ConnectedHeaders HandleAssumedValidHeader(ChainedHeader assumedValidHeader, ChainedHeader latestNewHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(assumedValidHeader), assumedValidHeader, nameof(latestNewHeader), latestNewHeader);
+
+            ChainedHeader bestTip = this.chainState.ConsensusTip;
+            ConnectedHeaders connectedHeaders = new ConnectedHeaders() {Consumed = latestNewHeader};
+
+            if (latestNewHeader.ChainWork > bestTip.ChainWork)
+            {
+                this.logger.LogDebug("Chained header '{0}' is the tip of a chain with more work than our current consensus tip.", latestNewHeader);
+
+                connectedHeaders = this.MarkBetterChainAsRequired(latestNewHeader);
+            }
+
+            this.MarkTrustedChainAsAssumedValid(assumedValidHeader);
+
+            this.logger.LogTrace("(-):{0}", connectedHeaders);
+            return connectedHeaders;
+        }
+
+        /// <summary>
+        /// When a header is before the last checkpoint it will be asked to be marked as AssumedValid and we'll ask to download it.
+        /// </summary>
+        private ConnectedHeaders HandleCheckpointsHeader(ChainedHeader chainedHeader, ChainedHeader latestNewHeader, CheckpointInfo checkpoint)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}',{4}.{5}:'{6}')", nameof(chainedHeader), chainedHeader, nameof(latestNewHeader), latestNewHeader, nameof(checkpoint), nameof(checkpoint.Hash), checkpoint.Hash);
+
+            if (checkpoint.Hash != chainedHeader.HashBlock)
+            {
+                this.logger.LogDebug("Chained header '{0}' does not match checkpoint '{1}'.", chainedHeader, checkpoint.Hash);
+                this.logger.LogTrace("(-)[INVALID_HEADER_NOT_MATCHING_CHECKPOINT]");
+                throw new InvalidHeaderException();
+            }
+
+            ChainedHeader subchainTip = chainedHeader;
+            if (chainedHeader.Height == this.checkpoints.GetLastCheckpointHeight())
+                subchainTip = latestNewHeader;
+
+            ConnectedHeaders connectedHeaders = this.MarkBetterChainAsRequired(subchainTip);
+            this.MarkTrustedChainAsAssumedValid(chainedHeader);
+
+            this.logger.LogTrace("(-):{0}", connectedHeaders);
+            return connectedHeaders;
+        }
+
+        /// <summary>
+        /// Check whether a header is in one of the following states
+        /// <see cref="BlockDataAvailabilityState.BlockAvailable"/>, <see cref="BlockDataAvailabilityState.BlockRequired"/>.
+        /// </summary>
+        private bool HeaderWasNotRequested(ChainedHeader chainedHeader)
```

Comment:
Maybe `HeaderWasNotRequested` => `HeaderWasRequested`? And invert the statement inside the method body. 
Not really insisting but imho it will be nice to have `!HeaderWasRequested()` instead of `HeaderWasNotRequested` that already has the negation.

Same for `HeaderWasNotMarkedAsValidated`

Reply:
Agree